File: balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # DFS, recursion
    def isBalanced(self, root: TreeNode) -> bool:
        height_balance = True
        
        def dfs(node=root):
            if node is None:
                return 0
            if not node.left and not node.right:
                return 1
            
            left = dfs(node.left)
            right = dfs(node.right)
            
            nonlocal height_balance
            if abs(left - right) > 1:
                height_balance = False
            
            return max(left, right) + 1
        
        dfs()
        return height_balance

File: balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    int maxDiff = 0;

    public boolean isBalanced(TreeNode root) {
        getHeight(root);
        return maxDiff <= 1? true: false;
    }
    public int getHeight(TreeNode node) {
        if(node == null)
            return 0;
        
        int left = getHeight(node.left) + 1;
        int right = getHeight(node.right) + 1;
        maxDiff = Math.max(maxDiff, Math.abs(left - right));
        return Math.max(left, right);
    }

    public boolean isBalanced(TreeNode root) {
        return check(root) != -1;
    }

    public int check(TreeNode node) {
        if(node == null)
            return 0;
        
        int left = check(node.left);
        int right = check(node.right);
        if(left == -1 || right == -1 || Math.abs(left - right) > 1)
            return -1;
        return Math.max(left, right) + 1;
    }
}